Add catalog_visibility param to products api

coderkevin opened the pull request:
Fixes #252

This adds a catalog_visibility=visible option to the product api which
will filter out any products that are excluded from the catalog.

#### Accessibility

No accessibility changes, this only filters products out in the preview.

### How to test the changes in this Pull Request:

1. Exclude some products from the catalog.
2. Try out each of the blocks and ensure those products don't show up in the previews.

coderkevin committed:
Add catalog_visibility param to products api

This adds a catalog_visibility=visible option to the product api which
will filter out any products that are excluded from the catalog.

joshuatf commented on includes/class-wgpb-products-controller.php:
```diff
@@ -230,10 +230,13 @@ public function prepare_item_for_response( $post, $request ) {
+		$attr_operator      = $request->get_param( 'attr_operator' );
+		$attributes         = $request->get_param( 'attributes' );
+		$tax_relation       = $request->get_param( 'tax_relation' );
+		$catalog_visibility = $request->get_param( 'catalog_visibility' );
```
I think we'll need to add an item to `get_collection_params()` for this param:

```php
		$params['catalog_visibility'] = array(
			'description'       => __( 'Determines if hidden or visible catalog products are shown.', 'woo-gutenberg-products-block' ),
			'type'              => 'string',
			'enum'              => array( 'visible', 'hidden' ),
			'sanitize_callback' => 'sanitize_text_field',
			'validate_callback' => 'rest_validate_request_arg',
		);
```

ryelle replied:
It would be good to support all visibility options – so the `enum` should be `array( 'visible', 'catalog', 'search', 'hidden' )` – you can see [the property in the item schema here.](https://github.com/woocommerce/woocommerce/blob/9ef6ea0be71a0719d72c5e511eb3c6b61330f183/includes/api/class-wc-rest-products-controller.php#L763-L769)

So `hidden` would get you products in both terms `'exclude-from-search'` & `'exclude-from-catalog'`,  `catalog` would be products `NOT IN 'exclude-from-catalog'`, and `search` would be `NOT IN 'exclude-from-search'`. `visible`, then, is `NOT IN `'exclude-from-search'` or `'exclude-from-catalog'`.

coderkevin replied:
Gotcha, thanks! I'll add code and tests.

ryelle replied:
`catalog_visibility` still needs to be added to `get_collection_params` here, the example from Josh above would work well (just update the enum).

joshuatf commented on includes/class-wgpb-products-controller.php:
```diff
@@ -259,6 +262,15 @@ protected function prepare_objects_query( $request ) {
+				'taxonomy' => 'product_visibility',
+				'field'    => 'term_id',
+				'terms'    => array( 'exclude-from-catalog' ),
+				'operator' => 'NOT IN',
```
Should we go ahead and add a `hidden` option that toggles the operator to `IN`?

joshuatf commented on assets/js/utils/get-query.js:
```diff
@@ -76,5 +76,8 @@ export default function getQuery( blockAttributes, name ) {
 	}
 
+	// Always only show visible products.
+	query.catalog_visibility = 'visible';
```
If we do go with the below additional `hidden` option I mentioned, we may want to add in an additional `catalogVisiblity` attribute and only set `query.catalog_visibility` if nothing is passed.

ryelle replied:
We probably won't need to control that from the block, so I don't think this is necessary.

ryelle replied:
Can you move this up to the initial settings around line 15? That's where we set up all the default values.

joshuatf reviewed (commented):
This is testing well and I'm only getting non-hidden products when querying with new param.

Left a couple minor comments and then I think this will be ready after adding in the `catalog_visibility` param to the getQuery tests in `assets/js/utils/test/get-query.js`.

ryelle commented on includes/class-wgpb-products-controller.php:
```diff
@@ -230,10 +230,13 @@ public function prepare_item_for_response( $post, $request ) {
+		$cat_operator       = $request->get_param( 'cat_operator' );
+		$attr_operator      = $request->get_param( 'attr_operator' );
+		$attributes         = $request->get_param( 'attributes' );
+		$tax_relation       = $request->get_param( 'tax_relation' );
```
Looks like there was a rebase issue here, `$attributes` & `$tax_relation` were removed in #401

coderkevin replied:
Thanks, I did this merge and must have just missed it.

ryelle reviewed (changes requested):
There's a rebase artifact left in here, and it would be good to see this support all the visibility options.

Can you also add a test case into the products-extras test file? You can set up a hidden product like this:

```php
$product = WC_Helper_Product::create_simple_product( false );
$product->set_catalog_visibility('hidden');
$product->save();
```

coderkevin committed:
Add product_visibility phpunit test

coderkevin committed:
Add catalog_visibility for catalog, search, hidden

coderkevin committed:
Fix util js test

The util js test checks for query parameters, so this adds the
category_visibility = 'visible' param.

ryelle reviewed (changes requested):
This is working great– I tried some API requests & they returned the expected visibility of products. The blocks are no longer previewing non-visible products.

I have two nitpicks, and can officially 👍  once those changes are made.

coderkevin committed:
Updated for comments

1. Added collection params
2. moved catalog_visibility default with others

coderkevin commented:
Updated for comments, ready for review

ryelle reviewed (approved):
Looks good, thanks for working on this 👍
